waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim

```diff
@@ -1024,23 +1052,58 @@ proc subscribeToGroupEvents*(ethClientUri: string,
   ## it collects all the events starting from the given `blockNumber`
   ## for every received event, it calls the `handler`
   let web3 = await newWeb3(ethClientUri)
-  var latestBlock: Quantity
+  let contract = web3.contractSender(MembershipContract, contractAddress)
+  
+  let blockTableRes = await getHistoricalEvents(ethClientUri, 
+                                                contractAddress, 
+                                                fromBlock=blockNumber)
+  if blockTableRes.isErr():
+    error "failed to retrieve historical events", error=blockTableRes.error()
+    return
+  # Update MT by batch
+  for blockNumber, members in blockTable.pairs():
+    debug "updating the Merkle tree", blockNumber=blockNumber, members=members
+    let res = handler(blockNumber, members)
+    if res.isErr():
+      error "failed to update the Merkle tree", error=res.error()
+
+  # We don't need the block table after this point
+  discard blockTable
+
+  var latestBlock: BlockNumber
+  let handleLog = proc(blockHeader: BlockHeader) {.async, gcsafe.} = 
+    try:
+      let membershipRegistrationLogs = await contract.getJsonLogs(MemberRegistered,
+                                                          blockHash = some(blockheader.hash))
+      if membershipRegistrationLogs.len == 0:
+        return
+      var members: seq[MembershipTuple]
+      for log in membershipRegistrationLogs:
+        let parsedEventRes = parse(MemberRegistered, log)
+        if parsedEventRes.isErr():
+          fatal "failed to parse the MemberRegistered event", error=parsedEventRes.error()
+          return
+        let parsedEvent = parsedEventRes.get()
+        members.add(parsedEvent)
+      let res = handler(blockHeader.number.uint, members)
+      if res.isErr():
+        error "failed to update the Merkle tree", error=res.error()
+    except CatchableError as err:
+      warn "failed to get logs", error=err.msg
+      return
   let newHeadCallback = proc (blockheader: BlockHeader) {.gcsafe.} =
-    latestBlock = blockheader.number
+    latestBlock = blockheader.number.uint
     debug "block received", blockNumber = latestBlock
+    # get logs from the last block
+    try:
+      asyncCheck handleLog(blockHeader)
```

Comment:
Up to this point, we only fetch historical events and then stop subscriptions to the further logs. What happened to the event subscription?

Reply:
Since we have to process blocks atomically, we cannot just subscribe to the event anymore. We have to subscribe to newHeads and then batch them according to blocks accordingly since the event subscription does not give us a list of events present in a block.

Reply:
makes sense, thanks!